### ISSUE TITLE
reverse DNS broken when query uses upper case

### ISSUE THREAD

**Issue by tobowers** [weaveworks/weave#2817]
Running weave 1.8.2, Docker 1.13.1, centos 7.2, aws, elastic container service

I noticed a problem while trying to cluster two rabbitmqs using an autocluster plugin and DNS. That plugin requires reverse-dns and I was able to isolate down to weave/erlang combination.

Forward DNS works fine in both docker/weave erlang consoles (only reverse DNS has a problem).

A docker-only (no weave) container will correctly resolve reverse dns from the erlang console:

```
Eshell V8.2.2  (abort with ^G)
1>  inet_res:gethostbyaddr({216,58,217,174}).
{ok,{hostent,"iad23s44-in-f14.1e100.net",[],inet,4,
             [{216,58,217,174}]}}
```

A container started through the weave proxy will *not* resolve reverse DNS.

```
Eshell V8.2.2  (abort with ^G)
1> inet_res:gethostbyaddr({216,58,217,174}).
{error,nxdomain}
```

However, bash on the container started through weave proxy *can* do the reverse DNS.

```
[root@ip-myip ~]# echo $DOCKER_HOST
unix:///var/run/weave/weave.sock
[root@ip-myip ~]# docker run -it --rm erlang /bin/bash -l
root@8974630db404:/# getent hosts 216.58.217.174
216.58.217.174  iad23s44-in-f174.1e100.net
```

I'm not familiar enough with weave/erlang internals to give this a proper diagnosis.

**Comment by bboreham** [comment]
@tobowers can you give the command-line used to start the container in the second example?

Since that doesn't look like a Weave address, the issue is in the recursive call to the upstream DNS.  `weave status` should say what the upstream DNS is set to; if that looks right then run Weave in debug mode, re-do the reverse DNS test and upload the logs (`docker logs weave`) please.

I.e. `weave stop`; `weave launch --log-level=debug`

**Comment by tobowers** [comment]
@bboreham thanks for looking. I discovered the problem using weave addresses but I thought the above google ip was going to be easier to diagnose. Here's an example using just weave addresses. It'll take me a sec to launch in debug mode... it's not as trivial as that command in my ECS-backed-by-systemd setup.

```
[root@ip-172-16-2-244 ~]# eval "$(weave env)"
[root@ip-172-16-2-244 ~]# docker run -it --rm erlang
Erlang/OTP 19 [erts-8.2.2] [source] [64-bit] [smp:2:2] [ds:2:2:10] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V8.2.2  (abort with ^G)
1> inet_res:lookup("elastic-master.weave.local", in, a).
[{10,32,0,2},{10,40,0,3},{10,40,0,2}]
2> inet_res:gethostbyaddr({10,32,0,2}).                 
{error,nxdomain}
3> inet_res:gethostbyaddr({10,40,0,3}).
{error,nxdomain}
4> inet_res:gethostbyaddr({10,40,0,2}).
{error,nxdomain}

```

**Comment by tobowers** [comment]
and here's the relevant logging around the DNS lookups from log-level debug

```
DEBU: 2017/02/27 10:46:08.736718 [nameserver 5a:4d:4f:c1:6d:56] lookup elastic-master.weave.local. -> &[10.32.0.2 10.40.0.2 10.40.0.3]
DEBU: 2017/02/27 10:46:08.736799 [nameserver 5a:4d:4f:c1:6d:56] response: ;; opcode: QUERY, status: NOERROR, id: 1
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;elastic-master.weave.local.	IN	 A

;; ANSWER SECTION:
elastic-master.weave.local.	1	IN	A	10.40.0.3
elastic-master.weave.local.	1	IN	A	10.40.0.2
elastic-master.weave.local.	1	IN	A	10.32.0.2

DEBU: 2017/02/27 10:46:13.792245 [nameserver 5a:4d:4f:c1:6d:56] reverse request: {MsgHdr:{Id:2 Response:false Opcode:0 Authoritative:false Truncated:false RecursionDesired:true RecursionAvailable:false Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:2.0.32.10.IN-ADDR.ARPA. Qtype:12 Qclass:1}] Answer:[] Ns:[] Extra:[]}
DEBU: 2017/02/27 10:46:13.792314 [nameserver 5a:4d:4f:c1:6d:56] response: ;; opcode: QUERY, status: NXDOMAIN, id: 2
;; flags: qr rd ra; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;2.0.32.10.IN-ADDR.ARPA.	IN	 PTR

```

**Comment by tobowers** [comment]
and a passing log from bash in same container (not erlang)

```
DEBU: 2017/02/27 10:48:20.161992 [nameserver 5a:4d:4f:c1:6d:56] reverse request: {MsgHdr:{Id:5031 Response:false Opcode:0 Authoritative:false Truncated:false RecursionDesired:true RecursionAvailable:false Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:2.0.32.10.in-addr.arpa. Qtype:12 Qclass:1}] Answer:[] Ns:[] Extra:[]}
DEBU: 2017/02/27 10:48:20.162049 [nameserver 5a:4d:4f:c1:6d:56] reverse lookup 10.32.0.2 -> elastic-master.weave.local.
DEBU: 2017/02/27 10:48:20.162098 [nameserver 5a:4d:4f:c1:6d:56] response: ;; opcode: QUERY, status: NOERROR, id: 5031
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0

;; QUESTION SECTION:
;2.0.32.10.in-addr.arpa.	IN	 PTR

;; ANSWER SECTION:
2.0.32.10.in-addr.arpa.	1	IN	PTR	elastic-master.weave.local.
```

**Comment by bboreham** [comment]
Thanks; can you run `weave status dns` on that same host where the failing lookup was run?

**Comment by tobowers** [comment]
```
# weave status dns
elastic-data 10.40.0.5       e4610530777e fa:e6:95:d9:5e:ca
elastic-master 10.32.0.2       8d657d1006b3 62:43:6c:0b:9a:1d
elastic-master 10.40.0.2       140f6966a10a fa:e6:95:d9:5e:ca
elastic-master 10.40.0.3       8a640f075a6b fa:e6:95:d9:5e:ca
servicequeue 10.46.0.1       83e06ade8956 5a:4d:4f:c1:6d:56
servicequeue 10.32.0.3       25307d54da0e 62:43:6c:0b:9a:1d
```

**Comment by tobowers** [comment]
huh... I'm just noticing the case difference in the question... 
failing:
```
Question:[{Name:2.0.32.10.IN-ADDR.ARPA. Qtype:12 Qclass:1}] Answer:[] Ns:[] Extra:[]}
```
passing:
```
Question:[{Name:2.0.32.10.in-addr.arpa. Qtype:12 Qclass:1}]
```

**Comment by bboreham** [comment]
Thanks; with enough logs all bugs are shallow :-) 
The most obvious thing is the case of the request:

    DEBU: 2017/02/27 10:46:13.792245 [nameserver 5a:4d:4f:c1:6d:56] reverse request: {MsgHdr:{Id:2 Response:false Opcode:0 Authoritative:false Truncated:false RecursionDesired:true RecursionAvailable:false Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:2.0.32.10.IN-ADDR.ARPA. Qtype:12 Qclass:1}] Answer:[] Ns:[] Extra:[]}

vs

    DEBU: 2017/02/27 10:48:20.161992 [nameserver 5a:4d:4f:c1:6d:56] reverse request: {MsgHdr:{Id:5031 Response:false Opcode:0 Authoritative:false Truncated:false RecursionDesired:true RecursionAvailable:false Zero:false AuthenticatedData:false CheckingDisabled:false Rcode:0} Compress:false Question:[{Name:2.0.32.10.in-addr.arpa. Qtype:12 Qclass:1}] Answer:[] Ns:[] Extra:[]}

**Comment by tobowers** [comment]
jinx. Right... but this is just generic erlang doing its thing.

**Comment by bboreham** [comment]
Yeah, it's a bug in WeaveDNS.  Probably here https://github.com/weaveworks/weave/blob/master/nameserver/dns.go#L225

**Comment by tobowers** [comment]
Ah ha... yeah that's definitely a problem.

**Comment by bboreham** [comment]
Fixed via 242b8907bfbc60cf358fe386e89801e8af4662fc
